### PR TITLE
[Feature] Truncate based on the sign of advantage after clipping

### PR DESCRIPTION
**Pull request by lehaoqu**
Add Dual-Clip PPO, which utilizes the `clip_ratio_c` to clip the ratio when the advantage is negative.
The Loss formular of Dual-Clip PPO  is following:

$$
loss = 
\begin{cases}
\min(\max(-A\times ratio, -A\times clip(ratio, 1-\epsilon_{low}, 1+\epsilon_{high})), -A\times Clip Ratio C ), & A <0\\
\max(-A\times ratio, -A\times clip(ratio, 1-\epsilon_{low}, 1+\epsilon_{high})), & A\geq 0
\end{cases}
$$

### CODE REVIEW

**Comment by garyzhang99** [comment]
Hi, thank you for your contribution!

I noticed a typo in the PR description formula; it should be $-A\times \text{Clip Ratio C}$ instead of $-A\times ratio\times \text{Clip Ratio C}$ in the first case.

Also, I think the functionality of this PR overlaps with #334. The implementations should be equivalent when setting `truncate_is_range_high == clip_ratio_c` in #334, assuming `1.0 + clip_range_high < truncate_is_range_high`.

Specifically:
- This PR: Applies a minimum bound only for negative advantages
- PR #334: Truncates the importance sampling ratio uniformly

When `truncate_is_range_high == clip_ratio_c` and the constraint above holds, both approaches achieve the same loss values.

Should we consider consolidating these two PRs?

**Comment by lehaoqu** [comment]
Yes @garyzhang99 . Besides, this PR decomposes the truncation of `pg_loss` **on the advantage dimension**, and compatibles [https://github.com/modelscope/Trinity-RFT/pull/334](https://github.com/modelscope/Trinity-RFT/pull/334). 
I convert the [https://github.com/modelscope/Trinity-RFT/pull/334](https://github.com/modelscope/Trinity-RFT/pull/334) into **clipping followed by truncating** based on [verl/core_algos.py](https://github.com/volcengine/verl/blob/main/verl/trainer/ppo/core_algos.py), and further truncate the `pg_loss` based on the different signs of advantage.

The [https://github.com/modelscope/Trinity-RFT/pull/334](https://github.com/modelscope/Trinity-RFT/pull/334) does truncating and clipping in the following order：
1. Truncate the `IS ratio`
2. Clip the `-A * IS ratio`

However, based on the implement of ppo loss in  [verl/core_algos.py](https://github.com/volcengine/verl/blob/main/verl/trainer/ppo/core_algos.py) as follows, we can find it first clips the `-A * IS ratio` and then truncates the `-A * IS ratio` when A is negative, to avoid the `-A * IS ratio` is too large. The order of verl ppo loss caculation as follows:
1. Clip the `-A * IS ratio`
2. Truncate the `-A * IS ratio`

```python
negative_approx_kl = log_prob - old_log_prob
# Clamp negative_approx_kl for stability
negative_approx_kl = torch.clamp(negative_approx_kl, min=-20.0, max=20.0)
ratio = torch.exp(negative_approx_kl)
ppo_kl = verl_F.masked_mean(-negative_approx_kl, response_mask)

#######################################
# 1. Clip the -A * ratio
pg_losses1 = -advantages * ratio
if cliprange_low is None:
    cliprange_low = cliprange
if cliprange_high is None:
    cliprange_high = cliprange
pg_losses2 = -advantages * torch.clamp(
    ratio, 1 - cliprange_low, 1 + cliprange_high
)  # - clip(ratio, 1-cliprange, 1+cliprange) * A
clip_pg_losses1 = torch.maximum(
    pg_losses1, pg_losses2
)  # max(-ratio * A, -clip(ratio, 1-cliprange, 1+cliprange) * A)
pg_clipfrac = verl_F.masked_mean(torch.gt(pg_losses2, pg_losses1).float(), response_mask)

#######################################
# 2. Truncate the -A * ratio
pg_losses3 = -advantages * clip_ratio_c
clip_pg_losses2 = torch.min(pg_losses3, clip_pg_losses1)
pg_clipfrac_lower = verl_F.masked_mean(
    torch.gt(clip_pg_losses1, pg_losses3) * (advantages < 0).float(), response_mask
)

pg_losses = torch.where(advantages < 0, clip_pg_losses2, clip_pg_losses1)
pg_loss = agg_loss(loss_mat=pg_losses, loss_mask=response_mask, loss_agg_mode=loss_agg_mode)
```

Since, to imporve the readability of code implement and decompose with different signs of advantage, I convert the [https://github.com/modelscope/Trinity-RFT/pull/334](https://github.com/modelscope/Trinity-RFT/pull/334) into **clipping followed by truncating**.

Notably, the formular of [https://github.com/modelscope/Trinity-RFT/pull/334](https://github.com/modelscope/Trinity-RFT/pull/334) is:

$$
\begin{aligned}
ratio' &= clip(ratio, trunc_{low}, trunc_{high})\\
loss &= \max(-A\times ratio', -A\times clip(ratio', 1-\epsilon_{low}, 1+\epsilon_{high}))
\end{aligned}
$$

Convert it to clipping followed by truncating.

$$
loss = \begin{cases}
\max(-A\times ratio, -A\times clip(ratio, trunc_{low}, \min(trunc_{high}, 1+\epsilon_{high}) )), & A>0 \\
\max(-A\times ratio, -A\times clip(ratio, \min(trunc_{low}, 1-\epsilon_{low}), trunc_{high} )), & A<0 \\
0, & A=0
\end{cases}
$$

**It is intuitive that $trunc_{high}>1+\epsilon_{high}$ and $trunc_{low}<1-\epsilon_{low}$**. The formula is simplified as follows:

$$
loss = \begin{cases}
\max(-A\times ratio, -A\times clip(ratio, trunc_{low}, 1+\epsilon_{high} )), & A>0 \\
\max(-A\times ratio, -A\times clip(ratio, 1-\epsilon_{low}, trunc_{high} )), & A<0 \\
0, & A=0
\end{cases}
$$

The above formula is equivalent to the following formula, which first clips `-A * IS ratio` based on $\epsilon$ and then truncates `-A * IS ratio`.

$$
loss = \begin{cases}
\min(\max(-A\times ratio, -A\times clip(ratio, 1-\epsilon_{low}, 1+\epsilon_{high} )), -A\times trunc_{low}), & A>0 \\
\min(\max(-A\times ratio, -A\times clip(ratio, 1-\epsilon_{low}, 1+\epsilon_{high} )), -A\times trunc_{high}), & A<0 \\
0, & A=0
\end{cases}
$$


**Since, the loss in [https://github.com/modelscope/Trinity-RFT/pull/334](https://github.com/modelscope/Trinity-RFT/pull/334) only depends on $trunc_{low}, \epsilon_{high}$ when advantage is positive, and depends on $trunc_{high}, \epsilon_{low}$  when advantage is negative.** The following image shows the IS ratio curve. 
[IS ratio of #334](https://p.ipic.vip/9sq5dm.jpg)

To further decompose on the advantage dimension, we need to the following two arguments to replace the `truncate_large_is`, which indicate whether to truncate advantages is positive and is negative.:  
- `truncate_adv_pos_is (bool)`  
- `truncate_adv_neg_is (bool)`  

Since, when we set `truncate_adv_pos_is=False`, we do not need to care the value of `truncate_is_range_low`; and when we set `truncate_adv_neg_is=False`, we also do not need to care the value of `truncate_is_range_high`

For example:
```
truncate_adv_pos_is: false
truncate_adv_neg_is: true
truncate_is_range_high: 2.0
clip_is_range_low: 0.2
clip_is_range_high: 0.2
```

```
truncate_adv_pos_is: true
truncate_adv_neg_is: false
truncate_is_range_low: 0.0
clip_is_range_low: 0.2
clip_is_range_high: 0.2
```

I have rebased #344 into this PR. The main implement of `pg_loss` on different signs of advantage as follows:
```python
# First clipping by clip_range, and calculate pg_clipfrac
pg_losses1 = -advantages * ratio
pg_losses2 = -advantages * torch.clamp(
    ratio, 1.0 - self.clip_range_low, 1.0 + self.clip_range_high  # type: ignore
)
pg_losses_clip = torch.maximum(pg_losses1, pg_losses2)
pg_clipfrac = masked_mean(torch.gt(pg_losses2, pg_losses1).float(), action_mask)

# Add IS truncation for positive advantages
if self.truncate_adv_pos_is:
    pg_losses_pos_trunc = -advantages * self.truncate_is_range_low
    pg_truncfrac_pos = masked_mean(
        torch.lt(pg_losses_pos_trunc, pg_losses_trunc) * (advantages > 0).float(),
        action_mask,
    )
    pg_losses_pos = torch.minimum(pg_losses_trunc, pg_losses_pos_trunc)
    pg_losses_trunc = torch.where(advantages > 0, pg_losses_pos, pg_losses_trunc)

# Add IS truncation for negative advantages
if self.truncate_adv_neg_is:
    pg_losses_neg_trunc = -advantages * self.truncate_is_range_high
    pg_truncfrac_neg = masked_mean(
        torch.lt(pg_losses_neg_trunc, pg_losses_trunc) * (advantages < 0).float(),
        action_mask,
    )
    pg_losses_neg = torch.minimum(pg_losses_trunc, pg_losses_neg_trunc)
    pg_losses_trunc = torch.where(advantages < 0, pg_losses_neg, pg_losses_trunc)

```

File: tests/algorithm/policy_loss_test.py [hunk 1]
```diff
@@ -108,3 +108,40 @@ def test_mix_policy_loss(self):
         self.assertTrue(torch.allclose(torch.tensor(metrics["usual/pg_loss"]), pg_loss))
         self.assertTrue(torch.allclose(torch.tensor(metrics["expert/sft_loss"]), sft_loss))
         self.assertTrue(torch.allclose(torch.tensor(metrics["loss"]), mix_loss))
+
+    def test_ppo_policy_loss_with_truncate_is(self):
```
Comment (garyzhang99) [inline tests/algorithm/policy_loss_test.py]: Consider adding a test case for when only `truncate_adv_neg_is` is set to true.

Comment (lehaoqu) [reply tests/algorithm/policy_loss_test.py]: Lastest commit has added `def test_ppo_policy_loss_with_truncate_adv_pos_is()` and `def test_ppo_policy_loss_with_truncate_adv_neg_is()` functions.

File: tests/algorithm/policy_loss_test.py [hunk 2]
```diff
@@ -108,3 +108,61 @@ def test_mix_policy_loss(self):
         self.assertTrue(torch.allclose(torch.tensor(metrics["usual/pg_loss"]), pg_loss))
         self.assertTrue(torch.allclose(torch.tensor(metrics["expert/sft_loss"]), sft_loss))
         self.assertTrue(torch.allclose(torch.tensor(metrics["loss"]), mix_loss))
+
+    def test_ppo_policy_loss_with_truncate_adv_pos_is(self):
```
Comment (garyzhang99) [inline tests/algorithm/policy_loss_test.py:112]: I recommend to keep the test for
```
policy_loss_fn_args["truncate_adv_neg_is"] = True
policy_loss_fn_args["truncate_adv_pos_is"] = True
```

Comment (lehaoqu) [reply tests/algorithm/policy_loss_test.py:112]: @garyzhang99  Ready!
Add a new test function: `test_ppo_policy_loss_with_truncate_adv_both_is()` to verify the decompose of truncation for different signs advantages.

When 
```
policy_loss_fn_args["truncate_adv_neg_is"] = True
policy_loss_fn_args["truncate_adv_pos_is"] = True
```
, this PR satisfies:
`ppo_loss_adv_both_truncated = ppo_loss_adv_pos_truncated + ppo_loss_adv_neg_truncated - ppo_loss_untruncated`

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 3]
```diff
@@ -34,6 +56,32 @@ def __init__(
...
+        self.truncate_adv_neg_is = truncate_adv_neg_is
+        if truncate_adv_pos_is:
+            self.truncate_is_range_low = truncate_is_range_low
+            assert (
+                self.truncate_is_range_low is not None
+            ), "truncate_is_range_low must be specified."
+            assert (
+                self.truncate_is_range_low >= 0.0
+            ), "truncate_is_range_low must be non-negative."
+            assert (self.truncate_is_range_low < 1.0-self.clip_range_low
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py:70]:
```suggestion
            assert (self.truncate_is_range_low < 1.0 - self.clip_range_low
```

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 4]
```diff
@@ -34,6 +56,32 @@ def __init__(
...
+            ), "truncate_is_range_low must be less than 1.0 - clip_range_low."
+        if truncate_adv_neg_is:
+            self.truncate_is_range_high = truncate_is_range_high
+            assert (
+                self.truncate_is_range_high is not None
+            ), "truncate_is_range_high must be specified."
+            assert (
+                self.truncate_is_range_high > 1.0+self.clip_range_high
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py:78]:
```suggestion
                self.truncate_is_range_high > 1.0 + self.clip_range_high
```

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 5]
```diff
@@ -46,25 +94,64 @@ def __call__(  # type: ignore
...
+        # This helps stabilize training when there are calculation discrepancies between
+        # rollout and training engines, especially for small probabilities
+        pg_truncfrac_pos, pg_truncfrac_neg = 0.0, 0.0
+        pg_losses_trunc = pg_losses_clip
+        
+        # Add IS truncation for positive advantages
+        if self.truncate_adv_pos_is:
+            pg_losses_pos_trunc = -advantages * self.truncate_is_range_low
+            pg_truncfrac_pos = masked_mean(
+                torch.lt(pg_losses_pos_trunc, pg_losses_trunc) * (advantages > 0).float(),
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py:115]:
```suggestion
                torch.lt(pg_losses_pos_trunc, pg_losses_trunc).float() * (advantages > 0),
```

File: trinity/algorithm/policy_loss_fn/ppo_policy_loss.py [hunk 6]
```diff
@@ -46,25 +94,64 @@ def __call__(  # type: ignore
...
+                action_mask,
+            )
+            pg_losses_pos = torch.minimum(pg_losses_trunc, pg_losses_pos_trunc)
+            pg_losses_trunc = torch.where(advantages > 0, pg_losses_pos, pg_losses_trunc)
+
+        # Add IS truncation for negative advantages
+        if self.truncate_adv_neg_is:
+            pg_losses_neg_trunc = -advantages * self.truncate_is_range_high
+            pg_truncfrac_neg = masked_mean(
+                torch.lt(pg_losses_neg_trunc, pg_losses_trunc) * (advantages < 0).float(),
```
Comment (yanxi-chen) [inline trinity/algorithm/policy_loss_fn/ppo_policy_loss.py:125]:
```suggestion
                torch.lt(pg_losses_neg_trunc, pg_losses_trunc).float() * (advantages < 0),
```

**Comment by yanxi-chen** [comment]
Thanks for contributing @lehaoqu! Just want to say that we might need to take this PR slow and make sure everything is perfect, since `ppo_policy_loss` plays such a central role in the whole Trinity system :)

**Comment by hiyuchang** [comment]
Thank you for your contribution! This PR is closed as it has been solved by #388 .
